contracts/ArchetypeLogic.sol

```diff
@@ -430,4 +408,49 @@ library ArchetypeLogic {
 
     return MerkleProofLib.verify(auth.proof, auth.key, keccak256(abi.encodePacked(account)));
   }
+
+  function getRandomTokenIds(
+    uint256[] memory tokenSupply,
+    uint32[] memory maxSupply,
+    uint32[] memory validIds,
+    uint256 quantity
+  ) public view returns (uint256[] memory) {
+    if (validIds.length == 0) {
+      validIds = new uint32[](maxSupply.length);
+      for (uint256 i = 0; i < validIds.length; i++) {
+        validIds[i] = uint32(i) + 1;
+      }
```

Comment:
Isn't this pricy to compute on each mint?

Reply:
Need to do some more testing, really depends on how many tokenIds are in the collection. Will be expensive for a large collection.

Reply:
But it is all in memory, so may not be too bad.

Reply:
see updated implementation @coindegen